README.md
[![Review Assignment Due Date](https://classroom.github.com/assets/deadline-readme-button-24ddc0f5d75046c5622901739e7c5dd533143b0c8e959d652212380cedb1ea36.svg)](https://classroom.github.com/a/DhYPBlwE)
# DO IT NOW

DO IT NOW  is a webapp in which users can list every task they have to do during their day. A user can make an account, log in and edit their personal informations. They also can add tasks to their to-do-list, edit it and mark it as completed.

## Table of content
- [DO IT NOW](#do-it-now)
  - [Table of content](#table-of-content)
  - [Installation](#installation)
    - [Requirements](#requirements)
      - [Step 1: Install Docker](#step-1-install-docker)
      - [Step 2: Verify Installation](#step-2-verify-installation)
    - [Configuration](#configuration)
      - [.env file](#env-file)
      - [Docker](#docker)
  - [Contributing](#contributing)
  - [License](#license)


## Installation

### Requirements
To run this project locally, you'll need to have Docker installed on your machine. Docker allows you to package and distribute applications along with their dependencies, ensuring consistent and reliable deployments.

If you haven't installed Docker yet, follow these steps to get started:

#### Step 1: Install Docker

1. **Windows**: Download the Docker Desktop installer from the official [Docker website](https://www.docker.com/products/docker-desktop) and follow the installation instructions.

2. **macOS**: Download the Docker Desktop for Mac installer from the official [Docker website](https://www.docker.com/products/docker-desktop) and follow the installation instructions.

3. **Linux**: Docker provides installation instructions for various Linux distributions. Visit the [Docker installation guide](https://docs.docker.com/engine/install/) for Linux to find instructions specific to your distribution.

#### Step 2: Verify Installation

After installing Docker, you can verify that it's working correctly by opening a terminal or command prompt and running the following command:

```sh
docker --version
docker-compose --version
```

### Configuration

#### .env file

Create a new .env file in the root directory of your project. This file will store sensitive information like database credentials.

Copy this in the .env file you have just created.

* `MYSQL_USERNAME ="username"`
* `MYSQL_ROOT_PASSWORD="password"`
* `MYSQL_DATABASE="database name"`

Replace username, your password, and your database name with your actual database details.

#### Docker

Run the following command to build and start the Docker containers defined in the docker-compose.yml file:

```bash
docker-compose up --build
```


## Contributing

Pull requests are welcome. For major changes, please open an issue first
to discuss what you would like to change.

## License
[MIT](https://choosealicense.com/licenses/mit/)



changelog.md
# changelog 


## [0.0.1] - 27-07-2023

### Added
    -added new mysql database with users table
    -added phpmysqladmin to manage database
    

## [0.0.2] - 30-07-2023

### Added 

    -rest api with crud for users
        - login endpoint
        - register endpoint
        - update user endpoint
        - delete user endpoint

## [0.0.3] - 31-07-2023µ

### fixed

    - response endpoints so that would send a json response

## [0.0.4] - 02-08-2023

### Added

    - added tasks table to database

## [0.0.5] - 03-08-2023

### Added

    - added tests for users endpoints

    - added tests for tasks endpoints

    - added crud for tasks
        - get tasks endpoint
        - create task endpoint
        - update task endpoint
        - delete task endpoint

## [0.0.6] - 07-08-2023

### added

    - frontend website with:
        - login page
        - register page
        - home page
        - profile page

## [0.0.7] - 07-08-2023

### updated

    - backend index file with all the new endpoints for tasks:
        - get tasks endpoint
        - create task endpoint
        - update task endpoint
        - delete task endpoint

images/backend/src/__tests__/task.test.js
const request = require('supertest');
const app = "localhost:3000"

beforeAll(async () => {
    // Create a new user
    const newUser ={ 
        firstName: 'John',
        lastName: 'Doe',
        email: 'john.doe@example.com',
        password: 'password123',
    };
    await request(app)
            .post('/createUser')
            .send(newUser);
});

afterAll(async () => {
    // Delete the user
    const user = {
        email: "john.doe@example.com",
        password: "password123"
    }
    
    await request(app)
            .delete('/deleteUser')
            .send(user);
})

describe('POST endpoint "/createTask"', () => {
    it('should create a new task for the user', async () => {
        const userEmail = 'john.doe@example.com';
        const userPassword = 'password123';
        const newTask = 'Buy fruits and vegetables for the week'

        // Make a request to the endpoint
        const response = await request(app)
            .post('/createTask')
            .send({ email: userEmail, password: userPassword, newTask });

        // Expectations
        expect(response.status).toBe(200);
        expect(response.body.status).toBe(200);
    });


    it('should return a status code of 401 if email is incorrect', async () => {
        const userEmail = 'fake@example.com';
        const userPassword = 'password123';
        const newTask = 'Buy fruits and vegetables for the week'

        // Make a request to the endpoint
        const response = await request(app)
            .post('/createTask')
            .send({ email: userEmail, password: userPassword, newTask });

        // Expectations
        expect(response.status).toBe(401);
        expect(response.body.status).toBe(401);
    });

    it('should return a status code of 401 if password is incorrect', async () => {
        const userEmail = 'john.doe@example.com';
        const userPassword = 'wrongpassword';
        const newTask = 'Buy fruits and vegetables for the week'

        // Make a request to the endpoint
        const response = await request(app)
            .post('/createTask')
            .send({ email: userEmail, password: userPassword, newTask });

        // Expectations
        expect(response.status).toBe(401);
        expect(response.body.status).toBe(401);
    });

});

describe('POST endpoint "/getTasks"', () => {
    it('should get tasks for a user based on the provided email and password', async () => {
        const userEmail = 'john.doe@example.com';
        const userPassword = 'password123';

        // Make a request to the endpoint
        const response = await request(app)
            .post('/getTasks')
            .send({ email: userEmail, password: userPassword });

        // Expectations
        expect(response.status).toBe(200);
        expect(response.body.status).toBe(200);
    });

    it('should return a status code of 401 if the user does not exist', async () => {
        const userEmail = 'nonexistent@example.com';
        const userPassword = 'password123';

        // Make a request to the endpoint
        const response = await request(app)
            .post('/getTasks')
            .send({ email: userEmail, password: userPassword });

        // Expectations
        expect(response.status).toBe(401);
        expect(response.body.status).toBe(401);
    });

    it('should return a status code of 401 if the password is incorrect', async () => {
        const userEmail = 'john.doe@example.com';
        const userPassword = 'wrongpassword';

        // Make a request to the endpoint
        const response = await request(app)
            .post('/getTasks')
            .send({ email: userEmail, password: userPassword });

        // Expectations
        expect(response.status).toBe(401);
        expect(response.body.status).toBe(401);
    });

    it('should return a status code of 401 if email or password are missing', async () => {
        // Make a request to the endpoint without providing the email and password parameters
        const response = await request(app).post('/getTasks');

        // Expectations
        expect(response.status).toBe(401);
        expect(response.body.status).toBe(401);
    });
});


describe('PUT endpoint "/updateTask"', () => {
    it('should update an existing task for the user', async () => {
        const userEmail = 'john.doe@example.com';
        const userPassword = 'password123';
        const currentTask= 'Buy fruits and vegetables for the week'
        const newTask = 'Updated task description'

        // Make a request to the endpoint
        const response = await request(app)
            .put(`/updateTask`)
            .send({ email: userEmail, password: userPassword, currentTask, newTask });

        // Expectations
        expect(response.status).toBe(200);
        expect(response.body.status).toBe(200);
    });


    it('should return a status code of 401 if the user does not exist', async () => {
        const userEmail = 'fake@example.com';
        const userPassword = 'password123';
        const currentTask = 'Updated task description'
        const newTask = 'Buy fruits and vegetables for the week'

        // Make a request to the endpoint
        const response = await request(app)
            .put(`/updateTask`)
            .send({ email: userEmail, password: userPassword, currentTask, newTask });

        // Expectations
        expect(response.status).toBe(401);
        expect(response.body.status).toBe(401);
    });

    it('should return a status code of 401 if the password is incorrect', async () => {
        const userEmail = 'john.doe@example.com';
        const userPassword = 'wrongpassword';
        const currentTask = 'Buy fruits and vegetables for the week'
        const newTask = 'Updated task description'

        // Make a request to the endpoint
        const response = await request(app)
            .put(`/updateTask`)
            .send({ email: userEmail, password: userPassword, currentTask, newTask });


        // Expectations
        expect(response.status).toBe(401);
        expect(response.body.status).toBe(401);
    })

    it('should return a status code of 400 if email or password are missing', async () => {
        const newTask = 'Buy fruits and vegetables for the week'
        const currentTask = 'Updated task description'

        // Make a request to the endpoint without providing the email and password parameters
        const response = await request(app)
            .put('/updateTask')
            .send({ currentTask, newTask });

        // Expectations
        expect(response.status).toBe(400);
        expect(response.body.status).toBe(400);
    })

});

describe('DELETE endpoint "/deleteTask"', () => {
    it('should delete an existing task for the user', async () => {
        const userEmail = 'john.doe@example.com';
        const userPassword = 'password123';
        const task = 'Buy fruits and vegetables for the week'; // Replace with the ID of the task to delete

        // Make a request to the endpoint
        const response = await request(app)
            .delete(`/deleteTask`)
            .send({ email: userEmail, password: userPassword, task });

        // Expectations
        expect(response.status).toBe(200);
        expect(response.body.status).toBe(200);
    });

    it('should return a status code of 401 if the user does not exist', async () => {
        const userEmail = 'fake@example.com';
        const userPassword = 'password123';
        const task = 'Buy fruits and vegetables for the week'; 

        // Make a request to the endpoint
        const response = await request(app)
            .delete(`/deleteTask`)
            .send({ email: userEmail, password: userPassword, task });


        // Expectations
        expect(response.status).toBe(401);
        expect(response.body.status).toBe(401);
    })

});


images/backend/src/index.js
import express from "express"
import bodyParser from "body-parser"
import cors from "cors"
import { databaseService } from "./services/databaseService.js"

const app = express()
app.use(cors())
app.use(express.static('./src/docs'))
app.use(bodyParser.urlencoded({
    extended: true
}))
app.use(bodyParser.json())
const port = 3000 || process.env.PORT

//get a message saying the server is running and on what port
app.listen(port, () => {
    console.log(`Example app listening at http://localhost:${port}`)
});

/**
 * object template
 * 
 * @typedef {object} User
 *      @property {string} firstName - first name of the user  
 *      @property {string} lastName - last name of the user
 *      @property {string} email - email of the user
 *      @property {string} password - password of the user
 * 
 * @typedef {object} userCredentials
 *      @property {string} email - email of the user
 *      @property {string} password - password of the user
 * 
 * @typedef {object} updatedUser
 *      @property {string} currentEmail - old email of user
 *      @property {string} newEmail - new email of user
 *      @property {string} currentPassword - old password of user
 *      @property {string} newPassword - new password of user
 * 
 * @typedef {object} updatedTask
 *      @property {string} currentTask - old task of user
 *      @property {string} newTask - new task of user
 */


/**GET endpoint, html page with all endpoints 
 * 
 * @returns {html} html page with all endpoints
*/
app.get('/', (request, response) => {
    response.status(200).redirect('./index.html')
})


/** POST endpoint, create a new user
 * 
 * @param {object} newUser - the new user to be created
 * @returns {string} - a message saying the user was created or an error message 
 */
app.post('/createUser', async (request, response) => {
    //check if all information is provided
    if (request.body.firstName && request.body.lastName && request.body.email && request.body.password) {
        try {
            const existingUserResult = await databaseService.getUser(request.body.email)
            //check if user already exists
            if (existingUserResult) {
                //return error message if user already exists
                response.status(400).send({
                    status: 400,
                    message: 'User already exists'
                })
                    
            } else {
                const newUser = {
                    firstName: request.body.firstName,
                    lastName: request.body.lastName,
                    email: request.body.email,
                    password: request.body.password
                }
                //add user to database
                await databaseService.addUser(newUser)
                response.status(200).send({
                    status: 200,
                    message: "User succesfully created"
                })
            }
        } catch (error) {
            //show error in console
            console.log(error);
            //send error message to frontend
            response.status(400).send({
                status: 400,
                message: error.message
            })
        }
    } else {
        //return error message if missing information
        response.status(401).send({
            status: 401,
            message: 'Missing information'
        })

    }
})


/**POST endpoint, check user credentials 
 * 
 * @param {object} userCredentials  - user credentials
 * @returns {string} - a message saying the user was logged in or an error message
 */
app.post('/login', async (request, response) => {
    //check if email and password are provided
    if (request.body.email && request.body.password) {
        try {
            const userCredentials = {
                email: request.body.email,
                password: request.body.password,
                newEmail: request.body.newEmail | 0,
                newPassword: request.body.newPassword | 0
            }
            //get user from database
            const user = await databaseService.getUser(userCredentials.email)
            //check if user exists and if password is correct
            if (user) {
                if (user.password === userCredentials.password) {
                    response.status(200).send({
                        status: 200,
                        message: "User succesfully logged in"
                    })
                } else {
                    response.status(401).send({
                        status: 401,
                        message: "incorrect credentials"
                    })
                }
            } else {
                response.status(401).send({
                    status: 401,
                    message: "incorrect credentials"
                })
            }
        } catch (error) {
            //show error in console
            console.log(error);
            //send error message in response
            response.status(401).send({
                status: 401,
                message: error.message
            })
        }
    } else {
        response.status(401).send({
            status: 401,
            message: 'incorrect credentials'
        })
    }
})


/** POST endpoint, get user data based on email
 * 
 * @param {object} userCredentials - user email and password
 * @returns {object} User - user data
 */
app.post('/getUserData', async (request, response) => {
    //check if email and password are provided
    if (request.body.email && request.body.password) {
        try {
            const email = request.body.email;
            // Assuming you have a function called getUserDataByEmail in your databaseService to fetch user data based on email
            const userData = await databaseService.getUser(email);
            //check if user exists and if password is correct
            //if user does not exist, return error message
            if (!userData) {
                return response.status(404).json({
                    status: 404,
                    message: "User data not found"
                });
            }
            //if password is incorrect, return error message
            if (userData.password !== request.body.password) {
                return response.status(401).json({
                    status: 401,
                    message: "Incorrect password"
                });
            } else {
                //if user exists and password is correct, return user data
                return response.status(200).json({
                    status: 200,
                    data: userData
                });
            }


        } catch (error) {
            // Handle the error here
            console.error(error);
            return response.status(500).json({
                status: 500, error:
                    "Internal server error"
            });
        }
    } else {
        //return error message if missing information
        return response.status(400).json({
            status: 400,
            error: "Missing information"
        });
    }

});

/**PUT endpoint, update user credentials 
 * 
 * @param {object} updatedUser - updated user credentials
 * @returns {string} - a message saying the user was updated or an error message
*/
app.put('/updateUser', async (request, response) => {
    console.log(request.body);

    //get updated user credentials from request body
    if (request.body.currentEmail && request.body.currentPassword) {
        let updatedUser = {
            currentEmail: request.body.currentEmail,
            currentPassword: request.body.currentPassword,
            newEmail: request.body.newEmail,
            newPassword: request.body.newPassword
        }

        try {
            let userCurrentCredentials = await databaseService.getUser(updatedUser.currentEmail)
            //retrieve user from database
            if (userCurrentCredentials) {
                //check if current password is correct
                if (userCurrentCredentials.password === updatedUser.currentPassword) {
                    //check if new email is not empty
                    databaseService.updateUser(updatedUser)
                    response.status(200).send({
                        status: 200,
                        message: "user succesfully updated"
                    })
                } else {

                    response.status(401).send({
                        status: 401,
                        message: "incorrect password"
                    })
                }
            } else {
                response.status(401).send({
                    status: 401,
                    message: "incorrect credentials"
                })
            }
        } catch (error) {
            // Handle the error here
            console.log(error);
            //send error back
            response.status(400).send({
                status: 400,
                message: error
            })
        }
    } else {
        response.status(400).send({
            status: 400,
            message: "missing information"
        })
    }
})



/**DELETE endpoint, delete user from database
 * 
 * @param {object} userCredentials - user credentials
 * @returns {string} - a message saying the user was deleted or an error message
*/
app.delete('/deleteUser', async (request, response) => {
    //get user credentials from request body
    if (request.body.email && request.body.password) {
        try {
            let userCredentials = {
                email: request.body.email,
                password: request.body.password
            }
            console.log(userCredentials);
            //get user from database
            let user = await databaseService.getUser(userCredentials.email)
            console.log(user);
            //check if user exists and if password is correct
            if (user) {
                if (user.password === userCredentials.password) {
                    //delete user from database
                    await databaseService.deleteUser(userCredentials.email)
                    response.status(200).send({
                        status: 200,
                        message: "User succesfully deleted"
                    })
                } else {
                    response.status(401).send({
                        status: 401,
                        message: "incorrect credentials"
                    })
                    console.log("log1");
                }
            } else {
                response.status(401).send({
                    status: 401,
                    message: "incorrect credentials"
                })
                console.log("log2");
            }
        } catch (error) {
            //show error in console
            console.log(error);
            //send error message in response
            response.status(400).send(error.message)
        }
    } else {
        response.status(401).send({
            status: 401,
            message: "incorrect credentials"
        })
    }
})

/**POST endpoint, create new task 
 * 
 * @param {object} userCredentials, newTask - user credentials and new task
*/
app.post('/createTask', async (request, response) => {
    //check if email and password are provided
    if (request.body.email && request.body.password && request.body.newTask) {
        try {
            //get user from database
            const result = await databaseService.getUser(request.body.email)
            //check if user exists and if password is correct
            if (result) {
                //check if user exists and if password is correct
                if (result.password === request.body.password) {
                    //create new task
                    await databaseService.createTask(result.id, request.body.newTask)
                    response.status(200).send({
                        status: 200,
                        message: "task succesfully created"
                    })
                } else {
                    //return error message if password is incorrect
                    response.status(401).send({
                        status: 401,
                        message: "incorrect credentials"
                    })
                }
            } else {
                //return error message if user does not exist
                response.status(401).send({
                    status: 401,
                    message: "missing details"
                })
            }
        } catch (error) {
            //show error in console
            console.log(error);
            //send error message in response
            response.status(400).send(error.message)
        }
        
            
    } else {
        response.status(401).send({
            status: 401,
            message: "incorrect credentials"
        })
    }
})



/**POST endpoint, get user tasks from database
 * 
 * @param {object} userCredentials - user credentials
 * @returns {object} - user tasks
*/
app.post('/getTasks', async (request, response) => {
    //check if email and password are provided
    if (request.body.email && request.body.password) {
        try {
            //get user from database
            const result = await databaseService.getUser(request.body.email)
            //check if user exists and if password is correct
            if (result) {
                //check if user exists and if password is correct
                if (result.password === request.body.password) {
                    //get tasks from database
                    const tasks = await databaseService.getTasks(result.id)
                    response.status(200).send({
                        status: 200,
                        tasks: tasks
                    })
                } else {
                    //return error message if password is incorrect
                    response.status(401).send({
                        status: 401,
                        message: "incorrect credentials"
                    })
                }
            } else {
                //return error message if user does not exist
                response.status(401).send({
                    status: 401,
                    message: "incorrect credentials"
                })
            }
        } catch (error) {
            //show error in console
            console.log(error);
            //send error message in response
            response.status(400).send(error.message)
        }
        
            
    } else {
        response.status(401).send({
            status: 401,
            message: "missing information"
        })
    }
})


/**PUT endpoint, update task
 * 
 * @param {object} userCredentials, updatedTask - user credentials and updated task
 * @returns {string} - a message saying the task was updated or an error message
*/
app.put('/updateTask', async (request, response) => {
    //check if email,password, currentTask and newTask are provided
    if (request.body.email && request.body.password && request.body.currentTask && request.body.newTask) {
        try {

            //get user from database
            const result = await databaseService.getUser(request.body.email)

            //check if user exists and if password is correct
            if (result) {
                //check if user exists and if password is correct
                if (result.password === request.body.password) {

                    const updatedTask = {
                        currentTask: request.body.currentTask,
                        newTask: request.body.newTask
                    }

                    //update task in database
                    await databaseService.updateTask(updatedTask)
                    response.status(200).send({
                        status: 200,
                        message: "task succesfully updated"
                    })
                } else {
                    //return error message if password is incorrect
                    response.status(401).send({
                        status: 401,
                        message: "incorrect credentials"
                    })
                }

            } else{
                //return error message if user does not exist
                response.status(401).send({
                    status: 401,
                    message: "incorrect credentials"
                })
            }
        } catch (error) {
            //show error in console
            console.log(error);
            //send error message in response
            response.status(400).send(error.message)
        }
    } else {
        response.status(400).send({
            status: 400,
            message: "missing information"
        })
    }
})


/**DELETE endpoint, delete task from database
 * 
 * @param {object} userCredentials, task - user credentials and task
 * @returns {string} - a message saying the task was deleted or an error message
*/
app.delete('/deleteTask', async (request, response) => {
    //check if email,password and task are provided
    if (request.body.email && request.body.password && request.body.task) {
        try {
            //get user from database
            const result = await databaseService.getUser(request.body.email)
            //check if user exists and if password is correct
            if (result) {
                //check if user exists and if password is correct
                if (result.password === request.body.password) {
                    //delete task from database
                    await databaseService.deleteTask(result.id,request.body.task)
                    response.status(200).send({
                        status: 200,
                        message: "task succesfully deleted"
                    })
                } else {
                    //return error message if password is incorrect
                    response.status(401).send({
                        status: 401,
                        message: "incorrect credentials"
                    })
                }
            } else {
                //return error message if user does not exist
                response.status(401).send({
                    status: 401,
                    message: "incorrect credentials"
                })
            }
        } catch (error) {
            //show error in console
            console.log(error);
            //send error message in response
            response.status(400).send(error.message)
        }
    } else {
        response.status(400).send({
            status: 400,
            message: "missing information"
        })
    }
})